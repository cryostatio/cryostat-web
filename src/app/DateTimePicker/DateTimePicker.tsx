/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TimePicker } from '@app/DateTimePicker/TimePicker';
import { useDayjs } from '@app/utils/hooks/useDayjs';
import { Timezone, defaultDatetimeFormat } from '@i18n/datetime';
import { isHourIn24hAM } from '@i18n/datetimeUtils';
import {
  ActionGroup,
  Bullseye,
  Button,
  CalendarMonth,
  Flex,
  FlexItem,
  Form,
  FormGroup,
  Tab,
  Tabs,
  TabTitleText,
  TextInput,
} from '@patternfly/react-core';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { TimezonePicker } from './TimezonePicker';

export interface DateTimePickerProps {
  onSelect?: (date: Date, timezone: Timezone) => void;
  prefilledDate?: Date; // timezone is ignored
  onDismiss: () => void;
}

type _TabKey = 'date' | 'time';

export const DateTimePicker: React.FC<DateTimePickerProps> = ({ onSelect, onDismiss, prefilledDate }) => {
  const { t } = useTranslation();
  const [dayjs, _] = useDayjs();
  const [activeTab, setActiveTab] = React.useState<_TabKey>('date');
  const [datetime, setDatetime] = React.useState<Date>(new Date());
  const [timezone, setTimezone] = React.useState<Timezone>(defaultDatetimeFormat.timeZone); // Not affected by user preferences

  const handleTabSelect = React.useCallback(
    (_: MouseEvent | React.MouseEvent, key: string | number) => setActiveTab(`${key}` as _TabKey),
    [setActiveTab],
  );

  const handleSubmit = React.useCallback(() => {
    onSelect && onSelect(datetime, timezone);
  }, [datetime, timezone, onSelect]);

  const handleCalendarSelect = React.useCallback(
    (_, date: Date) => {
      setDatetime((old) => {
        const wrappedOld = dayjs(old);
        return dayjs(date).hour(wrappedOld.hour()).minute(wrappedOld.minute()).second(wrappedOld.second()).toDate();
      });
      setActiveTab('time'); // Switch to time
    },
    [setDatetime, setActiveTab, dayjs],
  );

  const handleHourChange = React.useCallback(
    (hrIn24h: number) => {
      setDatetime((old) => dayjs(old).hour(hrIn24h).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleMinuteChange = React.useCallback(
    (m: number) => {
      setDatetime((old) => dayjs(old).minute(m).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleSecondChange = React.useCallback(
    (s: number) => {
      setDatetime((old) => dayjs(old).second(s).toDate());
    },
    [setDatetime, dayjs],
  );

  const handleMeridiemChange = React.useCallback(
    (isAM: boolean) => {
      setDatetime((old) => {
        const oldDayjs = dayjs(old);
        if (isAM !== isHourIn24hAM(oldDayjs.hour())) {
          return (isAM ? oldDayjs.subtract(12, 'hour') : oldDayjs.add(12, 'hour')).toDate();
        }
        return old;
      });
    },
    [setDatetime, dayjs],
  );

  const selectedDatetimeDisplay = React.useMemo(() => dayjs(datetime).format('L LTS'), [datetime, dayjs]);

  React.useEffect(() => {
    if (prefilledDate) {
      setDatetime(new Date(prefilledDate));
    }
  }, [setDatetime, prefilledDate]);

  return (
    <Form>
      <Tabs
        aria-label={t('DateTimePicker.ARIA_LABELS.TABS') || ''}
        onSelect={handleTabSelect}
        activeKey={activeTab}
        isFilled
        role="region"
      >
        <Tab key={'date'} eventKey={'date'} title={<TabTitleText>{t('DATE', { ns: 'common' })}</TabTitleText>}>
          <FormGroup>
            <Bullseye>
              <CalendarMonth
                className="datetime-picker__calendar"
                isDateFocused
                locale={dayjs.locale()}
                inlineProps={{
                  component: 'article',
                  ariaLabelledby: 'start-date',
                }}
                style={{ padding: 0 }}
                date={datetime}
                onChange={handleCalendarSelect}
              />
            </Bullseye>
          </FormGroup>
        </Tab>
        <Tab key={'time'} eventKey={'time'} title={<TabTitleText>{t('TIME', { ns: 'common' })}</TabTitleText>}>
          <FormGroup>
            <Bullseye>
              <TimePicker
                selected={{
                  hour24: dayjs(datetime).hour(), // 24hr format
                  minute: dayjs(datetime).minute(),
                  second: dayjs(datetime).second(),
                }}
                onHourSelect={handleHourChange}
                onMinuteSelect={handleMinuteChange}
                onSecondSelect={handleSecondChange}
                onMeridiemSelect={handleMeridiemChange}
              />
            </Bullseye>
          </FormGroup>
        </Tab>
      </Tabs>
      <FormGroup label={t('DateTimePicker.SELECTED_DATETIME')}>
        <Flex>
          <FlexItem>
            <TextInput
              id="selected-datetime"
              aria-label={t('DateTimePicker.ARIA_LABELS.DISPLAY_SELECTED_DATETIME') || ''}
              className="datetime-picker__datetime-selected-display"
              readOnly
              value={selectedDatetimeDisplay}
            />
          </FlexItem>
          <FlexItem>
            <TimezonePicker menuAppendTo={document.body} onTimezoneChange={setTimezone} selected={timezone} isCompact />
          </FlexItem>
        </Flex>
      </FormGroup>
      <ActionGroup style={{ marginTop: 0 }}>
        <Button variant="primary" onClick={handleSubmit}>
          {t('SELECT', { ns: 'common' })}
        </Button>
        <Button variant="link" onClick={onDismiss}>
          {t('CANCEL', { ns: 'common' })}
        </Button>
      </ActionGroup>
    </Form>
  );
};
