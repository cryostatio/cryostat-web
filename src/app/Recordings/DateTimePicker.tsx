/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


import {
  Button,
  ButtonVariant,
  DatePicker,
  Flex,
  FlexItem,
  InputGroup,
  isValidDate,
  Text,
  TimePicker,
  yyyyMMddFormat,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';

export interface DateTimePickerProps {
  onSubmit: (startDate) => void;
}

export const DateTimePicker: React.FunctionComponent<DateTimePickerProps> = (props) => {
  const [start, setStart] = React.useState(new Date(0));
  const [searchDisabled, setSearchDisabled] = React.useState(true);

  const onStartDateChange = React.useCallback(
    (inputDate, newStartDate) => {
      if (isValidDate(start) && isValidDate(newStartDate) && inputDate === yyyyMMddFormat(newStartDate)) {
        setStart(new Date(newStartDate));
        setSearchDisabled(false);
      } else {
        setSearchDisabled(true);
      }
    },
    [start, isValidDate, yyyyMMddFormat]
  );

  const onStartTimeChange = React.useCallback(
    (unused, hour, minute) => {
      let updated = new Date(start);
      updated.setUTCHours(hour, minute);
      setStart(updated);
    },
    [start, setStart, isValidDate]
  );

  const handleSubmit = React.useCallback(() => {
    props.onSubmit(`${start.toISOString()}`);
  }, [start, props.onSubmit]);

  return (
    <Flex>
      <FlexItem>
        <InputGroup>
          <DatePicker
            onChange={onStartDateChange}
            aria-label="Start date"
            placeholder="YYYY-MM-DD" />
          <TimePicker is24Hour aria-label="Start time" className="time-picker" onChange={onStartTimeChange} />
        </InputGroup>
        </FlexItem>
        <FlexItem>
          <Text>UTC</Text>
        </FlexItem>
        <FlexItem>
          <Button variant={ButtonVariant.control} aria-label="search button for start date" isDisabled={searchDisabled} onClick={handleSubmit}>
            <SearchIcon />
          </Button>
      </FlexItem>
    </Flex>
  );
};
