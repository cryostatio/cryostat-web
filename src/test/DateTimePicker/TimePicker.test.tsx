/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { format2Digit } from '@i18n/datetimeUtils';
import { cleanup, screen, within } from '@testing-library/react';
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { TimePicker } from '../../app/DateTimePicker/TimePicker';
import { renderDefault, testTranslate } from '../Common';

const onHourSelect = jest.fn((_) => undefined);
const onMinuteSelect = jest.fn((_) => undefined);
const onSecondSelect = jest.fn((_) => undefined);

const mockSelected = {
  hour24: 13,
  minute: 11,
  second: 11,
};

const mockSelectedIn12hr = {
  hour12: 1,
  minute: 11,
  second: 11,
  meridiem: 'PM',
};

describe('<TimePicker/>', () => {
  beforeEach(() => {
    jest.mocked(onHourSelect).mockReset();
    jest.mocked(onMinuteSelect).mockReset();
    jest.mocked(onSecondSelect).mockReset();
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <TimePicker
          selected={mockSelected}
          onHourSelect={onHourSelect}
          onMinuteSelect={onMinuteSelect}
          onSecondSelect={onSecondSelect}
        />
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly in 12hr mode', async () => {
    const { user } = renderDefault(
      <TimePicker
        selected={mockSelected}
        onHourSelect={onHourSelect}
        onMinuteSelect={onMinuteSelect}
        onSecondSelect={onSecondSelect}
      />
    );

    const switchenable24h = screen.getByLabelText(testTranslate('TimePicker.24HOUR'));
    expect(switchenable24h).toBeInTheDocument();
    expect(switchenable24h).toBeVisible();
    expect(switchenable24h).toBeChecked();

    await user.click(switchenable24h);

    const hInput = within(screen.getByLabelText(testTranslate('HOUR', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_HOUR12_VALUE')
    );
    expect(hInput).toBeInTheDocument();
    expect(hInput).toBeVisible();
    expect(hInput.getAttribute('value')).toBe(format2Digit(mockSelectedIn12hr.hour12));

    const mInput = within(screen.getByLabelText(testTranslate('MINUTE', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_MINUTE_VALUE')
    );
    expect(mInput).toBeInTheDocument();
    expect(mInput).toBeVisible();
    expect(mInput.getAttribute('value')).toBe(format2Digit(mockSelectedIn12hr.minute));

    const sInput = within(screen.getByLabelText(testTranslate('SECOND', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_SECOND_VALUE')
    );
    expect(sInput).toBeInTheDocument();
    expect(sInput).toBeVisible();
    expect(sInput.getAttribute('value')).toBe(format2Digit(mockSelectedIn12hr.second));

    const meridiem = screen.getByText(mockSelectedIn12hr.meridiem);
    expect(meridiem).toBeInTheDocument();
    expect(meridiem).toBeVisible();
    expect(meridiem.classList).toContain('selected');
  });

  it('should update time when increment/decrement buttons are clicked', async () => {
    const { user } = renderDefault(
      <TimePicker
        selected={mockSelected}
        onHourSelect={onHourSelect}
        onMinuteSelect={onMinuteSelect}
        onSecondSelect={onSecondSelect}
      />
    );

    const upHour = within(screen.getByLabelText(testTranslate('HOUR', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INCREMENT_HOUR24_VALUE')
    );
    expect(upHour).toBeInTheDocument();
    expect(upHour).toBeVisible();

    await user.click(upHour);

    expect(onHourSelect).toHaveBeenCalledTimes(1);
    expect(onHourSelect).toHaveBeenCalledWith(14);

    const downHour = within(screen.getByLabelText(testTranslate('HOUR', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.DECREMENT_HOUR24_VALUE')
    );
    expect(downHour).toBeInTheDocument();
    expect(downHour).toBeVisible();

    await user.click(downHour);

    expect(onHourSelect).toHaveBeenCalledTimes(2);
    expect(onHourSelect).toHaveBeenLastCalledWith(12);

    const upMinute = within(screen.getByLabelText(testTranslate('MINUTE', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INCREMENT_MINUTE_VALUE')
    );
    expect(upMinute).toBeInTheDocument();
    expect(upMinute).toBeVisible();

    await user.click(upMinute);

    expect(onMinuteSelect).toHaveBeenCalledTimes(1);
    expect(onMinuteSelect).toHaveBeenCalledWith(12);

    const downMinute = within(screen.getByLabelText(testTranslate('MINUTE', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.DECREMENT_MINUTE_VALUE')
    );
    expect(downMinute).toBeInTheDocument();
    expect(downMinute).toBeVisible();

    await user.click(downMinute);

    expect(onMinuteSelect).toHaveBeenCalledTimes(2);
    expect(onMinuteSelect).toHaveBeenCalledWith(10);

    const upSecond = within(screen.getByLabelText(testTranslate('SECOND', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INCREMENT_SECOND_VALUE')
    );
    expect(upSecond).toBeInTheDocument();
    expect(upSecond).toBeVisible();

    await user.click(upSecond);

    expect(onSecondSelect).toHaveBeenCalledTimes(1);
    expect(onSecondSelect).toHaveBeenCalledWith(12);

    const downSecond = within(screen.getByLabelText(testTranslate('SECOND', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.DECREMENT_SECOND_VALUE')
    );
    expect(downSecond).toBeInTheDocument();
    expect(downSecond).toBeVisible();

    await user.click(downSecond);

    expect(onSecondSelect).toHaveBeenCalledTimes(2);
    expect(onSecondSelect).toHaveBeenCalledWith(10);
  });

  it('should update time when time input is changed', async () => {
    const { user } = renderDefault(
      <TimePicker
        selected={mockSelected}
        onHourSelect={onHourSelect}
        onMinuteSelect={onMinuteSelect}
        onSecondSelect={onSecondSelect}
      />
    );

    const hInput = within(screen.getByLabelText(testTranslate('HOUR', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_HOUR24_VALUE')
    );
    expect(hInput).toBeInTheDocument();
    expect(hInput).toBeVisible();

    await user.type(hInput, '{Backspace}');
    expect(onHourSelect).toHaveBeenCalledTimes(1);
    expect(onHourSelect).toHaveBeenLastCalledWith(1);

    const mInput = within(screen.getByLabelText(testTranslate('MINUTE', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_MINUTE_VALUE')
    );
    expect(mInput).toBeInTheDocument();
    expect(mInput).toBeVisible();

    await user.type(mInput, '{Backspace}');
    expect(onHourSelect).toHaveBeenCalledTimes(1);
    expect(onHourSelect).toHaveBeenLastCalledWith(1);

    const sInput = within(screen.getByLabelText(testTranslate('SECOND', 'common'))).getByLabelText(
      testTranslate('TimeSpinner.INPUT_SECOND_VALUE')
    );
    expect(sInput).toBeInTheDocument();
    expect(sInput).toBeVisible();
    expect(sInput.getAttribute('value')).toBe(format2Digit(mockSelectedIn12hr.second));

    await user.type(sInput, '{Backspace}');
    expect(onHourSelect).toHaveBeenCalledTimes(1);
    expect(onHourSelect).toHaveBeenLastCalledWith(1);
  });
});
