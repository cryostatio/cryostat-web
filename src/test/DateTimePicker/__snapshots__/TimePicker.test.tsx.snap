// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePicker/> renders correctly 1`] = `
<div
  className="pf-v5-c-panel"
>
  <div
    className="pf-v5-c-panel__main"
    style={
      Object {
        "--pf-v5-c-panel__main--MaxHeight": undefined,
      }
    }
  >
    <div
      className="pf-v5-c-panel__main-body"
    >
      <div
        className="pf-v5-l-level pf-m-gutter"
      >
        <div>
          <div
            aria-label="Hour"
            className="pf-v5-l-stack"
          >
            <div
              className="pf-v5-l-stack__item"
            >
              <h4
                className="pf-v5-c-title pf-m-md datetime-picker__time-text-top-label"
                data-ouia-component-id="OUIA-Generated-Title-1"
                data-ouia-component-type="PF5/Title"
                data-ouia-safe={true}
              >
                Hour
              </h4>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Increment hour24 value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box up"
                data-ouia-component-id="OUIA-Generated-Button-primary-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <span
                className="pf-v5-c-form-control datetime-picker__number-input"
              >
                <input
                  aria-invalid={false}
                  aria-label="Input hour24 value"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                  data-ouia-component-type="PF5/TextInput"
                  data-ouia-safe={true}
                  disabled={false}
                  id="hour24-input"
                  max={23}
                  min={0}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="number"
                  value="13"
                />
              </span>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Decrement hour24 value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box down"
                data-ouia-component-id="OUIA-Generated-Button-primary-2"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            className="datetime-picker__colon-divider"
          >
            :
          </div>
        </div>
        <div>
          <div
            aria-label="Minute"
            className="pf-v5-l-stack"
          >
            <div
              className="pf-v5-l-stack__item"
            >
              <h4
                className="pf-v5-c-title pf-m-md datetime-picker__time-text-top-label"
                data-ouia-component-id="OUIA-Generated-Title-2"
                data-ouia-component-type="PF5/Title"
                data-ouia-safe={true}
              >
                Minute
              </h4>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Increment minute value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box up"
                data-ouia-component-id="OUIA-Generated-Button-primary-3"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <span
                className="pf-v5-c-form-control datetime-picker__number-input"
              >
                <input
                  aria-invalid={false}
                  aria-label="Input minute value"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                  data-ouia-component-type="PF5/TextInput"
                  data-ouia-safe={true}
                  disabled={false}
                  id="minute-input"
                  max={59}
                  min={0}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="number"
                  value="11"
                />
              </span>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Decrement minute value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box down"
                data-ouia-component-id="OUIA-Generated-Button-primary-4"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            className="datetime-picker__colon-divider"
          >
            :
          </div>
        </div>
        <div>
          <div
            aria-label="Second"
            className="pf-v5-l-stack"
          >
            <div
              className="pf-v5-l-stack__item"
            >
              <h4
                className="pf-v5-c-title pf-m-md datetime-picker__time-text-top-label"
                data-ouia-component-id="OUIA-Generated-Title-3"
                data-ouia-component-type="PF5/Title"
                data-ouia-safe={true}
              >
                Second
              </h4>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Increment second value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box up"
                data-ouia-component-id="OUIA-Generated-Button-primary-5"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <span
                className="pf-v5-c-form-control datetime-picker__number-input"
              >
                <input
                  aria-invalid={false}
                  aria-label="Input second value"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-3"
                  data-ouia-component-type="PF5/TextInput"
                  data-ouia-safe={true}
                  disabled={false}
                  id="second-input"
                  max={59}
                  min={0}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="number"
                  value="11"
                />
              </span>
            </div>
            <div
              className="pf-v5-l-stack__item"
            >
              <button
                aria-disabled={false}
                aria-label="Decrement second value"
                className="pf-v5-c-button pf-m-primary datetime-picker__time-spin-box down"
                data-ouia-component-id="OUIA-Generated-Button-primary-6"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-v5-c-icon pf-m-md"
                >
                  <span
                    className="pf-v5-c-icon__content"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      className="pf-v5-c-divider"
    />
    <div
      className="pf-v5-c-panel__footer"
    >
      <div
        className="pf-v5-c-helper-text"
      >
        <div
          className="pf-v5-c-helper-text__item"
        >
          <span
            className="pf-v5-c-helper-text__item-text"
          >
            Use 24-hour time
          </span>
        </div>
      </div>
      <label
        className="pf-v5-c-switch"
        data-ouia-component-id="OUIA-Generated-Switch-1"
        data-ouia-component-type="PF5/Switch"
        data-ouia-safe={true}
        htmlFor="24-hour-switch"
      >
        <input
          aria-label=""
          aria-labelledby="24-hour-switch-on"
          checked={true}
          className="pf-v5-c-switch__input"
          disabled={false}
          id="24-hour-switch"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="pf-v5-c-switch__toggle"
        />
        <span
          aria-hidden="true"
          className="pf-v5-c-switch__label pf-m-on"
          id="24-hour-switch-on"
        >
          24-hour
        </span>
        <span
          aria-hidden="true"
          className="pf-v5-c-switch__label pf-m-off"
          id="24-hour-switch-off"
        >
          24-hour
        </span>
      </label>
    </div>
  </div>
</div>
`;
